x/pubsub/keeper/msg_server.go

```diff
@@ -0,0 +1,360 @@
+	}
+
+	signer := msg.MustGetSigner()
+	signerAddress := signer.String()
+	if signerAddress != subscriber.Address {
+		return nil, sdkerrors.Wrapf(sdkerrors.ErrUnauthorized, "registered subscriber address must be signer: %s", subscriber.Address)
+	}
+
+	_, found = k.GetSubscriberIntent(ctx, subscriberIntent.SubscriptionId, subscriberAccAddress)
+	if found {
+		// TODO(bolten): would this UX be better if it removed the old one for you?
+		return nil, sdkerrors.Wrapf(types.ErrAlreadyExists, "subscriber already has intent for this subscription ID, must be removed first")
+	}
+
+	publisherIntent, found := k.GetPublisherIntent(ctx, subscriptionId, publisherDomain)
+	if !found {
+		return nil, sdkerrors.Wrapf(sdkerrors.ErrNotFound, "no publisher intent for domain %s and subscription ID %s found", publisherDomain, subscriptionId)
+	}
+
+	if publisherIntent.AllowedSubscribers == types.AllowedSubscribers_VALIDATORS {
+		// TODO(bolten): this implementation ends up making the module sort of non-generic but is necessary to allow
+		// orchestrator keys to manipulate subscriptions
+		var validatorI stakingtypes.ValidatorI
+		if validator := k.gravityKeeper.GetOrchestratorValidatorAddress(ctx, signer); validator == nil {
+			validatorI = k.stakingKeeper.Validator(ctx, sdk.ValAddress(signer))
+		} else {
+			validatorI = k.stakingKeeper.Validator(ctx, validator)
+		}
+
+		if validatorI == nil {
+			return nil, sdkerrors.Wrapf(sdkerrors.ErrUnauthorized, "publisher intent requires subscriber be a validator")
+		}
+	} else if publisherIntent.AllowedSubscribers == types.AllowedSubscribers_LIST {
+		// TODO(bolten): this is inefficient, perhaps could be replaced by a map?
+		found := false
+		for _, allowedAddress := range publisherIntent.AllowedAddresses {
+			if allowedAddress == signerAddress {
+				found = true
+				break
+			}
+		}
+
+		if !found {
+			return nil, sdkerrors.Wrapf(sdkerrors.ErrUnauthorized, "publisher intent requires subscriber to be in authorized list")
+		}
+	}
+
+	k.SetSubscriberIntent(ctx, subscriberAccAddress, *subscriberIntent)
+
+	ctx.EventManager().EmitEvents(
+		sdk.Events{
+			sdk.NewEvent(
+				sdk.EventTypeMessage,
+				sdk.NewAttribute(sdk.AttributeKeyModule, types.AttributeValueCategory),
+				sdk.NewAttribute(sdk.AttributeKeyAction, types.EventTypeAddSubscriberIntent),
+				sdk.NewAttribute(sdk.AttributeKeySender, msg.Signer),
+			),
+			sdk.NewEvent(
+				types.EventTypeAddSubscriberIntent,
+				sdk.NewAttribute(types.AttributeKeySubscriptionId, subscriberIntent.SubscriptionId),
+				sdk.NewAttribute(types.AttributeKeySubscriberAddress, subscriberIntent.SubscriberAddress),
+				sdk.NewAttribute(types.AttributeKeyPublisherDomain, subscriberIntent.PublisherDomain),
+			),
+		},
+	)
+
+	return &types.MsgAddSubscriberIntentResponse{}, nil
+}
+
+func (k Keeper) AddSubscriber(c context.Context, msg *types.MsgAddSubscriber) (*types.MsgAddSubscriberResponse, error) {
```

Comment:
With gas costs set to zero it's not really going to prevent spam. I think your point below about restricting these to where the sender is the subscriber address is important and missing. We could also customize this to require all subscribers to be validators, but that does sort of clash with the subscriber list functionality on the publisher side that is sort of meant to be generic. I don't think it's a huge deal if accounts register themselves as a subscriber, they won't be able to create an intent against a subscription ID requiring validators.